File: PAT_Basic_Level_by_lua/1118 如需挪车请致电/1118 如需挪车请致电.lua
-- 题目编号: 1118 如需挪车请致电
-- 实现原理: 解析11行输入，每行可能为拼音数字（直接映射）、sqrt开方运算或二元算术运算，计算结果取整后拼接输出。

-- 拼音数字到阿拉伯数字的映射
local w = {
    ling = 0, yi = 1, er = 2, san = 3, si = 4,
    wu = 5, liu = 6, qi = 7, ba = 8, jiu = 9
}
local o = {}  -- 存储每一位电话号码

-- 处理11位电话号码
for i = 1, 11 do
    local s = io.read("l")  -- 读取一行输入
    -- 尝试匹配二元算术运算：数字 + 运算符 + 数字
    local a, op, b = s:match("^(%d+)([+%-%*/%%%^])(%d+)$")
    local x  -- 存储计算结果
    
    if w[s] then
        -- 情况1：是拼音数字，直接映射
        x = w[s]
    elseif s:sub(1, 4) == "sqrt" then
        -- 情况2：是sqrt开方运算，对后面的数字开平方
        x = math.sqrt(tonumber(s:sub(5)))
    else
        -- 情况3：是二元算术运算，根据运算符计算
        a, b = tonumber(a), tonumber(b)
        if op == "+" then
            x = a + b
        elseif op == "-" then
            x = a - b
        elseif op == "*" then
            x = a * b
        elseif op == "/" then
            x = a / b
        elseif op == "%" then
            x = a % b
        else
            x = a ^ b  -- 幂运算
        end
    end
    
    -- 结果取整后加入电话号码
    o[#o + 1] = math.floor(x)
end

-- 拼接成完整的电话号码输出
print(table.concat(o))

File: PAT_Basic_Level_by_ruby/1118 如需挪车请致电/1118 如需挪车请致电.rb
# 1118 如需挪车请致电
#
# 题目描述：
#   ![phone.jpg](~/f29f986f-a871-4617-8fa7-e0cf4fe1d5d5.jpg)
#   上图转自新浪微博。车主用一系列简单计算给出了自己的电话号码，即：
#   2/2=1、3+2=5、\sqrt {9} = 3、\sqrt {9} = 3、0\% = 0、叁=3、5-2=3、9/3=3、1\times 3 = 3、2^3 = 8、8/2=4，最后得到的电话号码就是 153 3033 3384。
#   本题就请你写个程序自动完成电话号码的转换，以帮助那些不会计算的人。
#
# 实现原理：
#   根据车牌号和手机号的对应关系；查询车主的联系电话；使用哈希表快速查找
#

if __FILE__ == $PROGRAM_NAME
  names = %w[ling yi er san si wu liu qi ba jiu]
  results = []
  11.times do
    expression = gets.to_s.chomp
    if (match = expression.match(/\Asqrt(\d+)\z/))
      results << Math.sqrt(match[1].to_i).to_i
      next
    end

    if (match = expression.match(/\A(\d+)([+\-*\/%^])(\d+)\z/))
      left, op, right = match[1].to_i, match[2], match[3].to_i
      results << case op
                 when '+' then left + right
                 when '-' then left - right
                 when '*' then left * right
                 when '/' then left / right
                 when '%' then left % right
                 when '^' then left**right
                 end
    else
      results << names.index(expression)
    end
  end
  puts results.join
end
